craft_cli/dispatcher.py

```diff
@@ -0,0 +1,359 @@
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+"""Argument processing and command dispatching functionality."""
+
+import argparse
+from collections import namedtuple
+from typing import Any, Dict, List, Optional, Tuple, Type
+
+from craft_cli import EmitterMode, emit
+from craft_cli.errors import ArgumentParsingError, ProvideHelpException
+from craft_cli.helptexts import HelpBuilder
+
+# a helper to group some commands together
+CommandGroup = namedtuple("CommandGroup", "name commands")
+
+# global options: the name used internally, its type, short and long parameters, and help text
+GlobalArgument = namedtuple("GlobalArgument", "name type short_option long_option help_message")
+_DEFAULT_GLOBAL_ARGS = [
+    GlobalArgument(
+        "help",
+        "flag",
+        "-h",
+        "--help",
+        "Show this help message and exit",
+    ),
+    GlobalArgument(
+        "verbose",
+        "flag",
+        "-v",
+        "--verbose",
+        "Show debug information and be more verbose",
+    ),
+    GlobalArgument(
+        "quiet",
+        "flag",
+        "-q",
+        "--quiet",
+        "Only show warnings and errors, not progress",
+    ),
+    GlobalArgument(
+        "trace",
+        "flag",
+        "-t",
+        "--trace",
+        "Show all information needed to trace internal behaviour",
+    ),
+]
+
+
+class BaseCommand:
+    """Base class to build application commands.
+
+    Subclass this to create a new command; the subclass must define the following attributes:
+
+    - name: the identifier in the command line
+
+    - help_msg: a one line help for user documentation
+
+    - overview: a longer multi-line text with the whole command description
+
+    Also it may override the following ones to change their default:
+
+    - common: if it's a common/starter command, which are prioritized in the help (default to
+      False)
+
+    - needs_config: will ensure a config is provided when executing the command (default to False)
+
+    It also must/can override some methods for the proper command behaviour (see each
+    method's docstring).
+
+    The subclass must be declared in the corresponding section of main.COMMAND_GROUPS,
+    and will receive and store this group on instantiation (if overriding `__init__`, the
+    subclass must pass it through upwards).
+    """
+
+    common = False
+    needs_config = False
```

Comment:
I think I will remove this attribute from here.

If the commands needs a config or not is something very charmcraft specific, and it actually cannot be automatically verified here (charmcraft needs to check it!).